### PR TITLE
Declare external functions as pure/impure

### LINKED ISSUE #1478: Declare readTableData as impure
**Reported by hansolsson on 9 May 2014 12:48 UTC**
The functions readTableData in
Modelica.Blocks.Sources.CombiTimeTable
Modelica.Blocks.Tables.CombiTable1D
Modelica.Blocks.Tables.CombiTable1Ds
Modelica.Blocks.Tables.CombiTable2D
should be marked as impure (since it has an important side-effect) when MSL is based on Modelica Specification 3.3 (or later).

Note that it should not be done for MSL 3.2.x.

Milestone should probably be MSL 3.3; but it does not exist.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1478

### PR DESCRIPTION
**Pull request by beutlich**
Fix #1478.

@HansOlsson @MartinOtter Please review thoroughly.

For me, one open issue is the "dummy" table hack as for instance documented in https://github.com/modelica/ModelicaStandardLibrary/blob/ec70d1f10887371f11f14cb1abb696b60ee8cd6f/Modelica/Blocks/Tables.mo#L1320-L1330. This is a pure external function, but must not removed during optimization due to the intended side-effect to include the usertab header file. Is the recommendation to mark it as impure then?

### CODE REVIEW

**Comment by beutlich** [comment]
Another question is if the ctor/dtor functions of an ExternalObject should be declared pure or impure.

**Comment by beutlich** [comment]
@HansOlsson @MartinOtter Any comments at all?

**Comment by HansOlsson** [comment]
> Another question is if the ctor/dtor functions of an ExternalObject should be declared pure or impure.

I believe we need to clarify that in the specification (there is no special exception, but the examples don't use pure/impure for them). I would leave them without pure/impure for the moment.

Since the constructor returns newly allocated memory it doesn't return the same _value_ for the same arguments, which makes it impure - and the destructor deletes that memory, which seems like a side-effect (it's certainly not something that can be repeated or ignored). However, users don't call them directly - which could be seen as argument for not declaring them as either.

The argument for marking them as pure would be that the pointer in the external object isn't exposed in Modelica, and there may exist external object constructors that do more impure stuff. But I don't see any practical benefit, and as soon as there is one function that modifies the external object it's important to consider different external objects as distinct. 

However, it seems more important to have consistent rules for other functions on external objects. Since they don't return the external object it makes sense that functions modifying the visible state of the external object are marked as impure. (In C++ it would instead matter if the external object was sent as a normal pointer or as a pointer to const object.)

**Comment by HansOlsson** [comment]
> This is a pure external function, but must not removed during optimization due to the intended side-effect to include the usertab header file. Is the recommendation to mark it as impure then?

Not as written, as it's not legal to call an impure getUserTab as a binding for a non-parameter.

If you rewrite it as an initial equation/parameter binding then it would be possible to mark it impure; and I believe that rewriting it in that would anyway be a good change. 

As for "impure" the description of impurity doesn't consider source-code inclusion as for this, but if a function can be optimized away it seems logical that also includes its source-code; so it seems best to ensure that it isn't optimized away - and "impure" is one way.

Another possibility added in Modelica 3.4 would be to remove the output, mark it as pure, and put the call in an initial equation, since:

> A call to a function with no declared outputs is assumed to have desired side-effects or assertion checks. [A tool shall thus not remove such function calls, with exception of non-triggered assert calls. ...]

**Comment by HansOlsson** [comment]
As for external objects I realized that there is an exception for impure functions stating that they can be called from external object construction; so at least marking them as impure was considered.

**Comment by beutlich** [comment]
> It seems splitPathName should be marked as pure (or unmarked as it isn't external), since it's pure string manipulation.

splitPathName has the Impure annotation right now, which seems wrong then.

> Functions like getTableValue (etc) are not to be considered pure. You can call readTableData to change the internal state.

That is bad. In the cases for table data given as Modelica array or compiled as usertab, readTableData is useless and it is pure. I cases for table data read from file, it is almost always pure, except the exceptional case if readTableData was called some time. How to proceed?

> The random functions using state are pure as far as I know.

OK.

> If LU and the other lapack functions are pure or not can be debated (it will return different numbers if you compile and run on different machines).

I believe numerical issues are out of scope for pure/impure.

**Comment by beutlich** [comment]
> My idea is to declare readTableData as deprecated anyway.

@sjoelund OK for you to move readTableData to ObsoleteModelica to gain pure functions getTableValue  etc.?

**Comment by sjoelund** [comment]
Would the C-sources still be compatible with older MSL versions? If so, the only danger would be people loading ObsoleteModelica to ruin the purity.

**Comment by beutlich** [comment]
@casella @sjoelund Can I please get your review here 🙏 . Thanks.

**Comment by casella** [comment]
@beutlich, I share @MartinOtter concern that I'm not confident enough on the nuances of pure/impure semantics. I guess @sjoelund is a more competent reviewer, or maybe someone from other tool vendors, e.g. @svorkoetter or @eshmoylova

**Comment by MartinOtter** [comment]
> If `LU` and the other lapack functions are pure or not can be debated (it will return different numbers if you compile and run on different machines).

I do not understand this comment: All Lapack functions are pure functions (they will return exactly the same outputs, when called with the same inputs). The Modelica pure/impure keyword does not state anything about getting (slightly) different results on different machines. Otherwise, no external function would be pure.

**Comment by beutlich** [comment]
> 
> 
> > If `LU` and the other lapack functions are pure or not can be debated (it will return different numbers if you compile and run on different machines).
> 
> I do not understand this comment: All Lapack functions are pure functions (they will return exactly the same outputs, when called with the same inputs). The Modelica pure/impure keyword does not state anything about getting (slightly) different results on different machines. Otherwise, no external function would be pure.

Right. As I already replied, this is out of scope for this PR.

**Comment by beutlich** [comment]
@sjoelund @svorkoetter @christoff-buerger Can you please review. Thanks!

File: Modelica/Blocks/Tables.mo [hunk 1]
```diff
@@ -1310,35 +1262,34 @@ that can be used as a template for own developments. While &quot;usertab.c&quot;
...
-    annotation (Placement(transformation(extent={{-40,-34},{-10,-14}})));
-  Modelica.Blocks.Tables.CombiTable1Dv t_new(tableOnFile=true, tableName=\"TestTable_1D_a\")
+  parameter Real dummy(fixed=false) \"Dummy parameter\";
+  Modelica.Blocks.Tables.CombiTable1Dv table(tableOnFile=true, tableName=\"TestTable_1D_a\")
     annotation (Placement(transformation(extent={{-40,0},{-20,20}})));
   Modelica.Blocks.Sources.ContinuousClock clock
     annotation (Placement(transformation(extent={{-80,0},{-60,20}})));
 protected
-  encapsulated function getUsertab
+  encapsulated impure function getUsertab \"External dummy function to include \\\"usertab.c\\\"\"
```
Comment (sjoelund) [inline Modelica/Blocks/Tables.mo]: Why is usertab marked impure here? It is pure in this case, and if the function is not pure... does the rest of the tables work as it should?

File: Modelica/Math/package.mo [hunk 2]
```diff
@@ -1385,7 +1385,7 @@ has a unique solution.
 </html>"));
   end equalityLeastSquares;
 
-  function LU "LU decomposition of square or rectangular matrix"
+  pure function LU "LU decomposition of square or rectangular matrix"
```
Comment (sjoelund) [inline Modelica/Math/package.mo]: Not sure if this is pure or not. If it is translated on one machine and run on another, calls to LU during translation time might give very different results from runtime (because there are multiple solutions and there could be different lapack implementations involved; or the same one on different CPU architectures).

File: Modelica/Utilities/System.mo [hunk 3]
```diff
@@ -133,12 +128,11 @@ All returned values are of type Integer and have the following meaning:
 </html>"));
   end getTime;
 
-  function getPid "Retrieve the current process id"
+  impure function getPid "Retrieve the current process id"
```
Comment (sjoelund) [inline Modelica/Utilities/System.mo]: This is pure if it runs only during simulation time. Could use a comment that it is intended to stop the value from being calculated at compilation time.

File: Modelica/Math/package.mo (hunk 2 above)
Comment (beutlich) [reply Modelica/Math/package.mo]: Same argument as you remarked for getPid: This is pure if it runs only during simulation time.

File: Modelica/Blocks/Tables.mo (hunk 1 above)
Comment (beutlich) [reply Modelica/Blocks/Tables.mo]: Because it must not be removed during optimization. It is a makeshift solution to force the external Include annotation to get effective.

File: Modelica/Utilities/System.mo (hunk 3 above)
Comment (beutlich) [reply Modelica/Utilities/System.mo]: I guess you are right here since the PID does not change during simulation.

**Comment by beutlich** [comment]
@HansOlsson I changed getPid from impure to pure. OK for you, too?

**Comment by HansOlsson** [comment]
> @HansOlsson I changed getPid from impure to pure. OK for you, too?

Hmm... That's a complicated one, and I'm not sure why we have the function at all in MSL; I would assume it is used for some homebrewed parallel execution or process communication.

The problem is that a tool could attempt to constant-evaluate "getPID" (since it is a pure function with no non-literal arguments). That will generate an integer value, that doesn't work in the scenarios above.

For numeric routines, such as LU, similar optimizations may give slightly different results - but that error is small and bounded in some way, for getPid it is just another integer.

TL;DR: I believe getPid should be impure.

**Comment by beutlich** [comment]
> I'm not sure why we have the function at all in MSL

It was introduced in MSL v3.2.2 as some kind of autmoatic noise seed.

> TL;DR: I believe getPid should be impure.

OK, I removed it again and will leave it as impure (as is in MSL v3.2.2 and v3.2.3).

**Comment by beutlich** [comment]
I believe this PR is now ready to merge.

File: Modelica/Blocks/Tables.mo (hunk 1 above)
Comment (beutlich) [reply Modelica/Blocks/Tables.mo]: This was one of the main questions of this PR. See e.g. @HansOlsson's https://github.com/modelica/ModelicaStandardLibrary/pull/2858#issuecomment-496432733.
